A few bug fixes, and added a pair of new arguments

otgerp opened the pull request:
Changes:

- Fixed the generation failing if the _tsconfig.json_ file had an "_out_" option, as then Typescript aborted with: `Option 'noEmit' cannot be specified with option 'out'.`
  
- Fixed annotations in comments only being parsed for member properties, but not for e.g. top level classes (Fixes issue #37)
  
- Fixed bug with anonymous/inline types like the following example, where it was creating `Field` as `{ "$ref": "#/definitions/{ SubfieldA: number; }" }`
```
	interface TestObject {
		Field: {
			SubfieldA: number;
		}
	}
```

- Fixed crash with inline string literals like the one below. (this should fix issue #35)
```
	type SomeEnumType = "ValueA" | "ValueB" | "ValueC";
```
  
- Implemented support for string indexes (maps).
  For example, the following declaration will now generate a definition for `Field` with "additionalProperties" set to a reference to `SomeType`
```
	Field : { [id : string] : SomeType };
```
  
- Added an option to allow writing the output to a file: `-o file` / `--out file`

- Added the flag `--noExtraProps` to enforce by default all objects to have `"additionalProperties": false`. The flag is set to false by default.

YousefED commented:
Great stuff! Thank you so much.

Could you provide tests (should be fairly easy) for the new features supported? Mainly for top-level-comments, anonymous / inline types and string indices.

otgerp commented:
Some more changes (+ tests):

- Changed behavior for typescript `any`: it was being translated as `{ "type": "object" }`, now it 
 generates just an empty type spec `{}` (i.e. anything will match)
  
- Changed behavior for default value `null`: it was being ignored, but now it is set as `{ "default": null }`
 
- "integer" type can now be used in a typescript file by declaring `type integer = number;`, which will be recognized as a special type (the old way of `/** @type integer **/` still works also)

- Fixed bug with creating `"title`" for enums even if `--useTitle` was set to false
  
- Fixed various comments that still were being ignored (for this I had to reorder and move quite a lot of lines, that's why so many changes). Most of the cases should work now.

- Added flag `--aliasRefs` to enforce using `$ref` for typescript alias types (e.g. "`type Vec3 = number[];`"), instead of embedding the type. This effect is also enforced now for string literal types which create enums.

- Added tests `map-types` (string indexes), `comments` (comments with annotations at different places), `type-anonymous`, and `type-primitves` (for JSON schema primitive types with default values), and enabled and fixed the test `string-literals` (the expected json was wrong).

MarcusRiemer commented:
This immediatly solves three problems I am facing with the current "official" version, thanks a lot to both of you for your hard work!
